docs/design/architecture.md

```diff
@@ -0,0 +1,29 @@
+# Architecture 
+Kubernetes Console project consists of two main components. They are called here the
+frontend and the backend.
+
+The frontend is a single page web application that runs in a browser. It fetches all its
+business data from the backend using standard HTTP methods. It does not implement business logic;
+it only presents fetched data and sends requests to the backend for actions.
+
+The backend runs in a Kubernetes cluster as a kubernetes service. Alternatively, it may run anywhere
+outside of the cluster, given that it can connect to the master. The backend is a HTTP server that
+proxies data requests to appropriate remote backends (e.g., k8s apiserver or heapster) or implements
+business logic. The backend implements business logic when remote backends APIs do not
+support required use case directly, e.g., “get a list of pods with their CPU usage metric
+timeline”. Figure 1 outlines the architecture of the project.
+
+![Architecture Overview](architecture.png?raw=true "Architecture overview")
+
+*Figure 1: Project architecture overview*
+
+Rationale for having a backend that implements business logic:
+
+* Clear separation between the presentation layer (frontend) and business logic layer (backend).
+This is because every action goes through well defined API.
+* Transactional actions are easier to implement on the backend than on the frontend. Examples of
+such actions: "create a replication controller and a service for it" or "do a rolling update".
+* Possible code reuse from existing tools (e.g., kubectl) and upstream contributions to the tools.
+* Speed: getting composite data from backends is faster on the backend (if it runs close to the
+data sources). For example, getting a list of pods with their CPU utilization timeline
+requires at least two requests. Doing them on the backend shortens RTT.
```

Comment:
@liggitt Re stacking APIs: If all this was done on the browser frontend, would then it be okay? So, I postulate that this is no different from having a backend that exposes a private API for the browser frontend. Nobody else, except from the browser frontend, would be able to use the private API. This is just a matter of where to put the business logic: frontend or backend.

Creating a web UI that directly talks with "base" APIs is a very tempting way to go. We've put a great deal of thoughts on whether to have backend or be backend-less and the decision was hard to make. The general conclusion is that if the UI is read-only and accesses single API, then making is backen-less is OKay. Otherwise, tons of problems arise. From aforementioned transactions, through mixing business logic with views, slow browsers to ops issues (e.g., "how can I know whether this application even works?"). I've seen large UIs that started with backend-less approach which now, when the UIs are out in the wild, causes so much troubles. And of course, the backend-less approach didn't live for a long time because of the troubles.

Does this make sense @liggitt? Do you know any non-trivial UIs that are happy with being backend-less?